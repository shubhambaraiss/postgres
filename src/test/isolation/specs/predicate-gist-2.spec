# Test for page level predicate locking in gist
#
# Test to check false positives.
#
# Queries are written in such a way that an index scan(from one transaction) and an index insert(from another transaction) will try to access the different part(sub-tree) of the index.


setup
{
 create table gist_point_tbl(id int4, p point);
 create index gist_pointidx on gist_point_tbl using gist(p);
 insert into gist_point_tbl (id, p)
 select g, point(g*10, g*10) from generate_series(1, 1000) g;
}

teardown
{
 DROP TABLE gist_point_tbl;
}

session "s1"
setup		{ 
		  BEGIN ISOLATION LEVEL SERIALIZABLE;
		  set enable_seqscan=off;
		  set enable_bitmapscan=off;
		  set enable_indexonlyscan=on;
		}
step "rxy1"	{ select sum(p[0]) from gist_point_tbl where p >> point(6000,6000); }
step "wx1"	{ insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g; }
step "c1"	{ COMMIT; }

session "s2"
setup		{ 
		  BEGIN ISOLATION LEVEL SERIALIZABLE;
		  set enable_seqscan=off;
		  set enable_bitmapscan=off;
		  set enable_indexonlyscan=on;
		}

step "rxy2"	{ select sum(p[0]) from gist_point_tbl where p << point(1000,1000); }
step "wy2"	{ insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g; }
step "c2"	{ COMMIT; }
