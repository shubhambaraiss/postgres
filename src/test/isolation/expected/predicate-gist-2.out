Parsed test spec with 2 sessions

starting permutation: rxy1 wx1 c1 rxy2 wy2 c2
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c1: COMMIT;
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c2: COMMIT;

starting permutation: rxy1 wx1 rxy2 c1 wy2 c2
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step c1: COMMIT;
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c2: COMMIT;

starting permutation: rxy1 wx1 rxy2 wy2 c1 c2
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c1: COMMIT;
step c2: COMMIT;

starting permutation: rxy1 wx1 rxy2 wy2 c2 c1
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c2: COMMIT;
step c1: COMMIT;

starting permutation: rxy1 rxy2 wx1 c1 wy2 c2
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c1: COMMIT;
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c2: COMMIT;

starting permutation: rxy1 rxy2 wx1 wy2 c1 c2
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c1: COMMIT;
step c2: COMMIT;

starting permutation: rxy1 rxy2 wx1 wy2 c2 c1
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c2: COMMIT;
step c1: COMMIT;

starting permutation: rxy1 rxy2 wy2 wx1 c1 c2
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c1: COMMIT;
step c2: COMMIT;

starting permutation: rxy1 rxy2 wy2 wx1 c2 c1
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c2: COMMIT;
step c1: COMMIT;

starting permutation: rxy1 rxy2 wy2 c2 wx1 c1
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c2: COMMIT;
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c1: COMMIT;

starting permutation: rxy2 rxy1 wx1 c1 wy2 c2
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c1: COMMIT;
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c2: COMMIT;

starting permutation: rxy2 rxy1 wx1 wy2 c1 c2
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c1: COMMIT;
step c2: COMMIT;

starting permutation: rxy2 rxy1 wx1 wy2 c2 c1
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c2: COMMIT;
step c1: COMMIT;

starting permutation: rxy2 rxy1 wy2 wx1 c1 c2
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c1: COMMIT;
step c2: COMMIT;

starting permutation: rxy2 rxy1 wy2 wx1 c2 c1
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c2: COMMIT;
step c1: COMMIT;

starting permutation: rxy2 rxy1 wy2 c2 wx1 c1
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c2: COMMIT;
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c1: COMMIT;

starting permutation: rxy2 wy2 rxy1 wx1 c1 c2
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c1: COMMIT;
step c2: COMMIT;

starting permutation: rxy2 wy2 rxy1 wx1 c2 c1
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c2: COMMIT;
step c1: COMMIT;

starting permutation: rxy2 wy2 rxy1 c2 wx1 c1
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step c2: COMMIT;
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c1: COMMIT;

starting permutation: rxy2 wy2 c2 rxy1 wx1 c1
step rxy2: select sum(p[0]) from gist_point_tbl where p << point(1000,1000);
sum            

49500          
step wy2: insert into gist_point_tbl (id, p)
		  select g, point(g*50, g*50) from generate_series(1, 20) g;
step c2: COMMIT;
step rxy1: select sum(p[0]) from gist_point_tbl where p >> point(6000,6000);
sum            

3202000        
step wx1: insert into gist_point_tbl (id, p)
                  select g, point(g*500, g*500) from generate_series(12, 18) g;
step c1: COMMIT;
